### PR TITLE
refactor libpod config into libpod/config

### PR DESCRIPTION
**Pull request by vrothberg**

### CODE REVIEW

**Comment by baude** [comment]
code lgtm to me .. lets get the tests passing.

**Comment by vrothberg** [comment]
> code lgtm to me .. lets get the tests passing.

Found the needle in the haystack. The default config must _not_ set the `config.StorageConfig`, otherwise c/storage won't apply the correct labels to the created directories.

**Comment by vrothberg** [comment]
Now rebased. Still need to find the regression wrt to the container state.

**Comment by vrothberg** [comment]
@baude @rhatdan ... I can't believe we're (nearly) green. The on failing job is broken due to an old conmon version on the image.

File: libpod/config/config.go [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+type Config struct {
+	// NOTE: when changing this struct, make sure to update (*Config).Merge().
+
+	// StateType is the type of the backing state store. Avoid using multiple
+	// values for this with the same containers/storage configuration on the
+	// same system. Different state types do not interact, and each will see a
+	// separate set of containers, which may cause conflicts in
+	// containers/storage. As such this is not exposed via the config file.
+	//
+	// TODO: this is runtime data and doesn't belong here.
```
Comment (mheon) [inline libpod/config/config.go]: Eh? I do see this as part of config. If we ever added a SQL store, I'd want to be able to configure Podman to use it via libpod.conf

File: libpod/config/config.go [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+	// StorageConfig is the configuration used by containers/storage Not
+	// included in the on-disk config, use the dedicated containers/storage
+	// configuration file instead.
+	StorageConfig storage.StoreOptions `toml:"-"`
+
+	// StorageConfigRunRootSet indicates if the RunRoot has been explicitly set
+	// by the config or by the user. It's required to guarantee backwards
+	// compatibility with older versions of libpod for which we must query the
+	// database configuration. Not included in the on-disk config.
+	StorageConfigRunRootSet bool `toml:"-"`
```
Comment (mheon) [inline libpod/config/config.go]: Can we break this out to a sub-struct to group it together?

Comment (vrothberg) [reply libpod/config/config.go]: Sounds good to me.

File: libpod/config/config.go [hunk 3]
```diff
@@ -0,0 +1,606 @@
...
+	// older versions of libpod for which we must query the database
+	// configuration. Not included in the on-disk config.
+	TmpDirSet bool `toml:"-"`
+}
+
+// Config contains configuration options used to set up a libpod runtime
+type Config struct {
+	// NOTE: when changing this struct, make sure to update (*Config).Merge().
+
+	SetOptions // embedded struct to bundle the SetOptions
```
Comment (mheon) [inline libpod/config/config.go]: Maybe describe what the SetOptions are? Not obvious at first glance

File: libpod/config/config.go [hunk 4]
```diff
@@ -0,0 +1,606 @@
...
+	//DetachKeys is the sequence of keys used to detach a container.
+	DetachKeys string `toml:"detach_keys"`
+
+	// SDNotify tells Libpod to allow containers to notify the host systemd of
+	// readiness using the SD_NOTIFY mechanism.
+	SDNotify bool
+
+	// CgroupCheck verifies if the cgroup check for correct OCI runtime has been
+	// done.
+	CgroupCheck bool `toml:"cgroup_check,omitempty"`
```
Comment (mheon) [inline libpod/config/config.go]: IMO, let's nuke this. `runc` CGroups v2 support is almost ready, just pending merge of the devices controller. We probably will not be landing this in the 1.6 stable branch, so we have a good while until the next Podman release. Let's get rid of the crun/cgroups v2 hackery while we have a chance.

Comment (vrothberg) [reply libpod/config/config.go]: Same applies to forcing `crun` to be used when running on v2, right?

Comment (mheon) [reply libpod/config/config.go]: Yeah, let's git rid of as much special casing as we can.

**Pull request merged**